internal/app/connectconformance/test_case_library_test.go

```diff
@@ -901,6 +901,44 @@ func TestPopulateExpectedResponse(t *testing.T) {
 	})
 	require.NoError(t, err)
 
+	idempotentUnaryErrorJSONReqInfo, err := anypb.New(&conformancev1.ConformancePayload_RequestInfo{
+		RequestHeaders: requestHeaders,
+		Requests:       []*anypb.Any{unaryErrorReq},
+		ConnectGetInfo: &conformancev1.ConformancePayload_ConnectGetInfo{
+			QueryParams: []*conformancev1.Header{
+				{
+					Name: "message",
+					// unaryErrorReq marshaled as json
+					Value: []string{"{\"responseDefinition\":{\"responseHeaders\":[{\"name\":\"fooHeader\",\"value\":[\"fooHeaderVal\"]},{\"name\":\"barHeader\",\"value\":[\"barHeaderVal1\",\"barHeaderVal2\"]}],\"error\":{\"code\":8,\"message\":\"all resources exhausted\"},\"responseTrailers\":[{\"name\":\"fooTrailer\",\"value\":[\"fooTrailerVal\"]},{\"name\":\"barTrailer\",\"value\":[\"barTrailerVal1\",\"barTrailerVal2\"]}]}}"},
+				},
+				{
+					Name:  "encoding",
+					Value: []string{"json"},
+				},
+			},
+		},
+	})
+	require.NoError(t, err)
+
+	idempotentUnaryErrorProtoReqInfo, err := anypb.New(&conformancev1.ConformancePayload_RequestInfo{
+		RequestHeaders: requestHeaders,
+		Requests:       []*anypb.Any{unaryErrorReq},
+		ConnectGetInfo: &conformancev1.ConformancePayload_ConnectGetInfo{
+			QueryParams: []*conformancev1.Header{
+				{
+					Name: "message",
+					// unaryErrorReq marshaled and base64 encoded
+					Value: []string{"Cq4BChkKCWZvb0hlYWRlchIMZm9vSGVhZGVyVmFsCikKCWJhckhlYWRlchINYmFySGVhZGVyVmFsMRINYmFySGVhZGVyVmFsMiIbCgpmb29UcmFpbGVyEg1mb29UcmFpbGVyVmFsIiwKCmJhclRyYWlsZXISDmJhclRyYWlsZXJWYWwxEg5iYXJUcmFpbGVyVmFsMhobCAgSF2FsbCByZXNvdXJjZXMgZXhoYXVzdGVk"},
```

Comment:
This and the JSON-serialized message above are likely best done by creating a message literal and then serializing it. Otherise, it's not clear how one updates the test case if something needs to be tweaked (like a field here or there).

Reply:
If most of the definition is defined using JSON (like suggested above), you could leave out parts like that from the JSON and also supply the message instances, and then serialize them and store them inside the reuqest/response definition inside the loop that processes the test cases.